app/src/main/java/com/alphawallet/app/walletconnect/WCClient.kt

```diff
@@ -0,0 +1,313 @@
+                id = id,
+                result = result
+        )
+        return encryptAndSend(gson.toJson(response))
+    }
+
+    fun rejectRequest(id: Long, message: String = "Rejected by the user"): Boolean {
+        val response = JsonRpcErrorResponse(
+                id = id,
+                error = JsonRpcError.serverError(
+                        message = message
+                )
+        )
+        return encryptAndSend(gson.toJson(response))
+    }
+
+    private fun decryptMessage(text: String): String {
+        val message = gson.fromJson<WCSocketMessage>(text)
+        val encrypted = gson.fromJson<WCEncryptionPayload>(message.payload)
+        val session = this.session
+                ?: throw IllegalStateException("Session is null")
+        return String(WCCipher.decrypt(encrypted, session.key.toByteArray()), Charsets.UTF_8)
+    }
+
+    private fun invalidParams(id: Long): Boolean {
+        val response = JsonRpcErrorResponse(
+                id = id,
+                error = JsonRpcError.invalidParams(
+                        message = "Invalid parameters"
+                )
+        )
+        return encryptAndSend(gson.toJson(response))
+    }
+
+    private fun handleMessage(payload: String) {
+        try {
+            val request = gson.fromJson<JsonRpcRequest<JsonArray>>(payload, typeToken<JsonRpcRequest<JsonArray>>())
+            val method = request.method
+            if (method != null) {
+                handleRequest(request)
+            } else {
+                onCustomRequest(request.id, payload)
+            }
+        } catch (e: InvalidJsonRpcParamsException) {
+            invalidParams(e.requestId)
+        }
+    }
+
+    private fun handleRequest(request: JsonRpcRequest<JsonArray>) {
+        when (request.method) {
+            WCMethod.SESSION_REQUEST -> {
+                val param = gson.fromJson<List<WCSessionRequest>>(request.params)
+                        .firstOrNull() ?: throw InvalidJsonRpcParamsException(request.id)
+                handshakeId = request.id
+                remotePeerId = param.peerId
+                onSessionRequest(request.id, param.peerMeta)
+            }
+            WCMethod.SESSION_UPDATE -> {
+                val param = gson.fromJson<List<WCSessionUpdate>>(request.params)
+                        .firstOrNull() ?: throw InvalidJsonRpcParamsException(request.id)
+                if (!param.approved) {
+                    killSession()
+                }
+            }
+            WCMethod.ETH_SIGN -> {
+                val params = gson.fromJson<List<String>>(request.params)
+                if (params.size < 2)
+                    throw InvalidJsonRpcParamsException(request.id)
+                onEthSign(request.id, WCEthereumSignMessage(params, WCEthereumSignMessage.WCSignType.MESSAGE))
+            }
+            WCMethod.ETH_PERSONAL_SIGN -> {
+                val params = gson.fromJson<List<String>>(request.params)
```

Comment:
We are already including a gson library. Can we use the one we're already importing (via a Java call) rather than adding a duplicate library?

Reply:
Are you referring to the kotson library? It is not a duplicate gson lib, just a lib containing helper/extension functions

Reply:
When I see unknown libraries which have people's names in the path I get nervous about them. What does it do? Is it an open source library, is it credible? I'll have to check the library out.

Reply:
Ahh I see what you mean. No worries, I'll remove it and use the regular way!

Reply:
I checked it out it looks legit with a lot of contributors and is mainly for code clarity. If you haven't started the change then keep it as it is.